File: attribute-parser-compiler/src/main/java/pyxis/uzuki/live/attribute/parser/compiler/AttributeParserProcessor.java
package pyxis.uzuki.live.attribute.parser.compiler;

import com.google.auto.service.AutoService;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Filer;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;

import pyxis.uzuki.live.attribute.parser.annotation.AttributeParser;
import pyxis.uzuki.live.attribute.parser.annotation.CustomView;
import pyxis.uzuki.live.attribute.parser.compiler.holder.CustomViewHolder;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrBooleanModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrColorModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrDimensionModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrDimensionPixelSizeModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrDrawableModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrFloatModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrIntModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrResourceModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.AttrStringModel;
import pyxis.uzuki.live.attribute.parser.compiler.model.BaseAttrModel;
import pyxis.uzuki.live.attribute.parser.compiler.utils.Utils;

@SupportedSourceVersion(SourceVersion.RELEASE_8)
@AutoService(javax.annotation.processing.Processor.class)
public class AttributeParserProcessor extends AbstractProcessor {
    private List<CustomViewHolder> mCustomViewHolderList = new ArrayList<>();
    private Map<String, List<AttrIntModel>> mAttrIntMap = new HashMap<>();
    private Map<String, List<AttrBooleanModel>> mAttrBooleanMap = new HashMap<>();
    private Map<String, List<AttrColorModel>> mAttrColorMap = new HashMap<>();
    private Map<String, List<AttrDimensionModel>> mAttrDimensionMap = new HashMap<>();
    private Map<String, List<AttrDimensionPixelSizeModel>> mAttrDimensionPixelSizeMap = new HashMap<>();
    private Map<String, List<AttrDrawableModel>> mAttrDrawableMap = new HashMap<>();
    private Map<String, List<AttrFloatModel>> mAttrFloatMap = new HashMap<>();
    private Map<String, List<AttrResourceModel>> mAttrResourceMap = new HashMap<>();
    private Map<String, List<AttrStringModel>> mAttrStringMap = new HashMap<>();
    private Filer mFiler;
    private String mPackageName;

    @Override
    public synchronized void init(ProcessingEnvironment env) {
        super.init(env);
        mFiler = env.getFiler();
    }

    @Override
    public boolean process(Set<? extends TypeElement> set, RoundEnvironment roundEnvironment) {
        processAnnotation(roundEnvironment);
        for (CustomViewHolder customViewHolder : mCustomViewHolderList) {
            writeAttributes(customViewHolder);
        }
        return true;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Utils.getSupportedAnnotationSet();
    }

    private void processAnnotation(RoundEnvironment env) {
        for (Element element : env.getElementsAnnotatedWith(AttributeParser.class)) {
            AttributeParser parser = element.getAnnotation(AttributeParser.class);
            mPackageName = parser.value();
        }

        for (Element element : env.getElementsAnnotatedWith(CustomView.class)) {
            mCustomViewHolderList.add(new CustomViewHolder(ClassName.get((TypeElement) element)));
        }

        Utils.findAnnotatedWith(env, Utils.getAttrIntPair(), mAttrIntMap);
        Utils.findAnnotatedWith(env, Utils.getAttrStringPair(), mAttrStringMap);
        Utils.findAnnotatedWith(env, Utils.getAttrBooleanPair(), mAttrBooleanMap);
        Utils.findAnnotatedWith(env, Utils.getAttrColorPair(), mAttrColorMap);
        Utils.findAnnotatedWith(env, Utils.getAttrDimensionPair(), mAttrDimensionMap);
        Utils.findAnnotatedWith(env, Utils.getAttrDimensionPixelSizePair(), mAttrDimensionPixelSizeMap);
        Utils.findAnnotatedWith(env, Utils.getAttrDrawablePair(), mAttrDrawableMap);
        Utils.findAnnotatedWith(env, Utils.getAttrFloatPair(), mAttrFloatMap);
        Utils.findAnnotatedWith(env, Utils.getAttrResourcePair(), mAttrResourceMap);
    }

    private void writeAttributes(CustomViewHolder customViewHolder) {
        TypeSpec.Builder builder = TypeSpec.classBuilder(customViewHolder.getSimpleName() + Constants.ATTRIBUTES)
                .addModifiers(Modifier.PUBLIC);

        List<BaseAttrModel> models = Utils.getModelList(customViewHolder.getSimpleName(), mAttrBooleanMap,
                mAttrColorMap, mAttrDimensionMap, mAttrDrawableMap, mAttrIntMap, mAttrDimensionPixelSizeMap,
                mAttrFloatMap, mAttrResourceMap, mAttrStringMap);

        for (BaseAttrModel model : models) {
            builder.addField(createAttrsFieldSpec(model));
        }

        builder.addField(createRFieldSpec());
        builder.addMethod(createObtainApplyMethodSpec(customViewHolder));
        builder.addMethod(createApplyMethodSpec(customViewHolder, models));
        builder.addMethod(createPrintVariableMethodSpec(customViewHolder, models));
        builder.addMethod(createBindAttributesMethodSpec(customViewHolder, models));

        JavaFile javaFile = JavaFile.builder(Constants.PACKAGE_NAME, builder.build()).build();

        try {
            javaFile.writeTo(System.out);
            javaFile.writeTo(mFiler);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private FieldSpec createRFieldSpec() {
        TypeName typeName = ClassName.bestGuess(mPackageName + Constants.R_FILE);
        return FieldSpec.builder(typeName, Constants.R_VARIABLE, Modifier.PRIVATE).build();
    }

    private FieldSpec createAttrsFieldSpec(BaseAttrModel model) {
        String variableName = model.getAnnotatedElementName();
        TypeName typeName = Utils.bestGuess(model.getAnnotatedElementClass());
        return FieldSpec.builder(typeName, variableName, Modifier.PRIVATE, Modifier.STATIC).build();
    }

    private MethodSpec createPrintVariableMethodSpec(CustomViewHolder customViewHolder, List<BaseAttrModel> models) {
        MethodSpec.Builder builder = MethodSpec.methodBuilder(Constants.PRINT_VARIABLES)
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC);

        StringBuilder variablesBuilder = new StringBuilder();
        int maxinum = 0;
        for (int i = 0; i < models.size(); i++) {
            BaseAttrModel model = models.get(i);
            String variableName = model.getAnnotatedElementName();
            String className = model.getAnnotatedElementClass();
            String line = "\"\\n" + className + " " + variableName + " = \" + " + variableName + " +  \n";

            variablesBuilder.append(line);

            int count = line.length();
            maxinum = Math.max(maxinum, count);
        }

        StringBuilder stringBuilder = new StringBuilder();
        String lastLine = Utils.multiply("=", maxinum);
        String firstLine = Utils.multiply("=", (maxinum - customViewHolder.getSimpleName().length() - 2) / 2);
        String firstLineMessage = String.format("\"%s %s %s\" + \n", firstLine, customViewHolder.getSimpleName(), firstLine);

        stringBuilder.append(firstLineMessage);
        stringBuilder.append(variablesBuilder.toString());
        stringBuilder.append(String.format("\"\\n%s\"", lastLine));

        String message = stringBuilder.toString();
        builder.addCode(String.format(Constants.STATEMENT_LOG, customViewHolder.getSimpleName(), message));

        return builder.build();
    }

    private MethodSpec createObtainApplyMethodSpec(CustomViewHolder customViewHolder) {
        TypeName classTypeName = customViewHolder.getClassName();
        String classTypeParameterName = customViewHolder.getSimpleName().substring(0, 1).toLowerCase() +
                customViewHolder.getSimpleName().substring(1);

        MethodSpec.Builder builder = Utils.getMethodSpec(Constants.APPLY, Modifier.PUBLIC, Modifier.STATIC)
                .addParameter(classTypeName, classTypeParameterName)
                .addParameter(Constants.ATTRIBUTE_SET_CLASS_NAME, Constants.SET)
                .addCode(String.format(Constants.STATEMENT_OBTAIN_APPLY, classTypeParameterName,
                        classTypeParameterName, customViewHolder.getSimpleName()));

        return builder.build();
    }

    private MethodSpec createApplyMethodSpec(CustomViewHolder customViewHolder, List<BaseAttrModel> models) {
        TypeName classTypeName = customViewHolder.getClassName();
        String classTypeParameterName = customViewHolder.getSimpleName().substring(0, 1).toLowerCase() +
                customViewHolder.getSimpleName().substring(1);

        MethodSpec.Builder builder = Utils.getMethodSpec(Constants.APPLY, Modifier.PUBLIC, Modifier.STATIC)
                .addParameter(classTypeName, classTypeParameterName)
                .addParameter(Constants.TYPED_ARRAY_CLASS_NAME, Constants.ARRAY)
                .addCode(Constants.BIND_ATTRIBUTES_INVOKE);

        for (BaseAttrModel model : models) {
            String variableName = model.getAnnotatedElementName();

            builder.addCode(String.format(Constants.STATEMENT_APPLY, classTypeParameterName,
                    variableName, variableName));
        }

        return builder.build();
    }

    private MethodSpec createBindAttributesMethodSpec(CustomViewHolder customViewHolder, List<BaseAttrModel> models) {
        MethodSpec.Builder builder = Utils.getMethodSpec(Constants.BIND_ATTRIBUTES, Modifier.PRIVATE, Modifier.STATIC)
                .addParameter(Constants.TYPED_ARRAY_CLASS_NAME, Constants.ARRAY)
                .addCode(Constants.STATEMENT_BINDATTRIBUTES);

        for (BaseAttrModel model : models) {
            Utils.addCode(builder, model, customViewHolder.getSimpleName());
        }

        builder.addCode(Constants.STATEMENT_RECYCLE);

        return builder.build();
    }
}

File: attribute-parser-compiler/src/main/java/pyxis/uzuki/live/attribute/parser/compiler/AttributeParserProcessor.kt
package pyxis.uzuki.live.attribute.parser.compiler

import com.google.auto.service.AutoService
import com.squareup.kotlinpoet.*
import pyxis.uzuki.live.attribute.parser.annotation.AttributeParser
import pyxis.uzuki.live.attribute.parser.annotation.CustomView
import pyxis.uzuki.live.attribute.parser.compiler.holder.CustomViewHolder
import pyxis.uzuki.live.attribute.parser.compiler.model.*
import pyxis.uzuki.live.attribute.parser.compiler.utils.*
import java.io.File
import java.util.*
import javax.annotation.processing.AbstractProcessor
import javax.annotation.processing.Processor
import javax.annotation.processing.RoundEnvironment
import javax.annotation.processing.SupportedSourceVersion
import javax.lang.model.SourceVersion
import javax.lang.model.element.TypeElement

@SupportedSourceVersion(SourceVersion.RELEASE_8)
@AutoService(Processor::class)
class AttributeParserProcessor : AbstractProcessor() {
    private val mCustomViewHolderList = ArrayList<CustomViewHolder>()
    private val mAttrIntMap = HashMap<String, List<AttrIntModel>>()
    private val mAttrBooleanMap = HashMap<String, List<AttrBooleanModel>>()
    private val mAttrColorMap = HashMap<String, List<AttrColorModel>>()
    private val mAttrDimensionMap = HashMap<String, List<AttrDimensionModel>>()
    private val mAttrIntegerMap = HashMap<String, List<AttrIntegerModel>>()
    private val mAttrFractionMap = HashMap<String, List<AttrFractionModel>>()
    private val mAttrFloatMap = HashMap<String, List<AttrFloatModel>>()
    private val mAttrResourceMap = HashMap<String, List<AttrReferenceModel>>()
    private val mAttrStringMap = HashMap<String, List<AttrStringModel>>()
    private var mPackageName: String? = null

    override fun process(set: Set<TypeElement>, roundEnvironment: RoundEnvironment): Boolean {
        processAnnotation(roundEnvironment)
        for (customViewHolder in mCustomViewHolderList) {
            writeAttributes(customViewHolder)
        }
        return true
    }

    override fun getSupportedAnnotationTypes() = supportedAnnotationSet

    private fun processAnnotation(env: RoundEnvironment) {
        env.getElementsAnnotatedWith(AttributeParser::class.java)
                .map { it.getAnnotation(AttributeParser::class.java) }
                .forEach { mPackageName = it.value }

        env.getElementsAnnotatedWith(CustomView::class.java).mapTo(mCustomViewHolderList) {
            CustomViewHolder((it as TypeElement).asClassName(), processingEnv.elementUtils.getPackageOf(it).toString())
        }

        env.run {
            findAnnotatedWith(attrIntPair, mAttrIntMap)
            findAnnotatedWith(attrStringPair, mAttrStringMap)
            findAnnotatedWith(attrBooleanPair, mAttrBooleanMap)
            findAnnotatedWith(attrColorPair, mAttrColorMap)
            findAnnotatedWith(attrDimensionPair, mAttrDimensionMap)
            findAnnotatedWith(attrIntegerPair, mAttrIntegerMap)
            findAnnotatedWith(attrFractionPair, mAttrFractionMap)
            findAnnotatedWith(attrFloatPair, mAttrFloatMap)
            findAnnotatedWith(attrResourcePair, mAttrResourceMap)
        }
    }

    private fun writeAttributes(holder: CustomViewHolder) {
        val classTypeName = holder.className
        val classTypeParameterName = holder.simpleName.substring(0, 1).toLowerCase() + holder.simpleName.substring(1)
        val simpleName = holder.simpleName
        val fileName = simpleName + Constants.ATTRIBUTES

        val builder = TypeSpec.objectBuilder(fileName)

        val models = getModelList(simpleName, mAttrBooleanMap, mAttrColorMap, mAttrDimensionMap,
                mAttrIntegerMap, mAttrIntMap, mAttrFractionMap, mAttrFloatMap, mAttrResourceMap, mAttrStringMap)

        for (model in models) {
            builder.addProperty(createAttrsFieldSpec(model))
        }

        builder.addProperty(createRFieldSpec())
        builder.addFunction(createObtainApplyMethodSpec(classTypeName, classTypeParameterName, holder.simpleName))
        builder.addFunction(createApplyMethodSpec(classTypeName, classTypeParameterName, models))
        builder.addFunction(createPrintVariableMethodSpec(simpleName, models))
        builder.addFunction(createBindAttributesMethodSpec(simpleName, models))

        val kaptKotlinGeneratedDir = processingEnv.options[KAPT_KOTLIN_GENERATED_OPTION_NAME]
                ?.replace("kaptKotlin", "kapt")
                ?.let { File(it, "$fileName.kt") }
                ?: throw IllegalArgumentException("No output directory")

        val typeSpec = builder.build()
        val fileSpec = FileSpec.builder(holder.packageName, fileName).addType(typeSpec).build()
        fileSpec.writeTo(kaptKotlinGeneratedDir)
    }

    private fun createRFieldSpec(): PropertySpec {
        val typeName = ClassName.bestGuess(mPackageName!! + Constants.R_FILE)
        return PropertySpec.builder(Constants.R_VARIABLE, typeName, KModifier.PRIVATE).initializer("R()").build()
    }

    private fun createAttrsFieldSpec(model: BaseAttrModel): PropertySpec {
        val variableName = model.annotatedElementName
        val typeName = model.annotatedElementClass.bestGuess()
        return PropertySpec.builder(variableName, typeName, KModifier.PRIVATE)
                .mutable(true).initializer(typeName.initializer()).build()
    }

    private fun createPrintVariableMethodSpec(simpleName: String, models: List<BaseAttrModel>): FunSpec {
        val builder = Constants.PRINT_VARIABLES.getFunSpec(KModifier.PUBLIC)
                .addAnnotation(JvmStatic::class)
        val variablesBuilder = StringBuilder()
        var maxinum = 0
        for (model in models) {
            val variableName = model.annotatedElementName
            val className = model.annotatedElementClass
            val line = "\"\\n$className $variableName = \" + $variableName +  \n"
            variablesBuilder.append(line)
            maxinum = Math.max(maxinum, line.length)
        }

        val stringBuilder = StringBuilder()
        val lastLine = multiply(Constants.EQUALS, maxinum)
        val firstLine = multiply(Constants.EQUALS, (maxinum - simpleName.length - 2) / 2)
        val firstLineMessage = String.format(Constants.STATEMENT_LOG_FIRST_LINE, firstLine, simpleName, firstLine)

        stringBuilder.append(firstLineMessage)
        stringBuilder.append(variablesBuilder.toString())
        stringBuilder.append(String.format(Constants.STATEMENT_LOG_LAST_LINE, lastLine))

        val message = stringBuilder.toString()
        builder.addStatement("%T.d(%S, %L)", Constants.LOG_CLASS, simpleName, message)

        return builder.build()
    }

    private fun createObtainApplyMethodSpec(classTypeName: TypeName, classTypeParameterName: String, simpleName: String): FunSpec {
        val builder = Constants.APPLY.getFunSpec(KModifier.PUBLIC)
                .addParameter(classTypeParameterName, classTypeName)
                .addParameter(Constants.SET, Constants.ATTRIBUTE_SET_CLASS_NAME)
                .addAnnotation(JvmStatic::class)
                .addCode(Constants.STATEMENT_OBTAIN_APPLY.format(classTypeParameterName, classTypeParameterName, simpleName))

        return builder.build()
    }

    private fun createApplyMethodSpec(classTypeName: TypeName, classTypeParameterName: String, models: List<BaseAttrModel>): FunSpec {
        val builder = Constants.APPLY.getFunSpec(KModifier.PUBLIC)
                .addParameter(classTypeParameterName, classTypeName)
                .addParameter(Constants.ARRAY, Constants.TYPED_ARRAY_CLASS_NAME)
                .addAnnotation(JvmStatic::class)
                .addCode(Constants.BIND_ATTRIBUTES_INVOKE)

        models.map { it.annotatedElementName }
                .forEach {
                    builder.addCode("%L.%L = %L;\n", classTypeParameterName, it, it)
                }

        return builder.build()
    }

    private fun createBindAttributesMethodSpec(simpleName: String, models: List<BaseAttrModel>): FunSpec {
        val builder = Constants.BIND_ATTRIBUTES.getFunSpec(KModifier.PRIVATE)
                .addParameter(Constants.ARRAY, Constants.TYPED_ARRAY_CLASS_NAME)
                .addCode(Constants.STATEMENT_BINDATTRIBUTES)

        for (model in models) {
            builder.addCode(model, simpleName)
        }

        builder.addCode(Constants.STATEMENT_RECYCLE)

        return builder.build()
    }

    companion object {
        const val KAPT_KOTLIN_GENERATED_OPTION_NAME = "kapt.kotlin.generated"
    }
}